src/node/byzantine_identity.h

```diff
@@ -0,0 +1,342 @@
+
+  class ByzantineIdentitiesHook : public kv::ConsensusHook
+  {
+    kv::Version version;
+    std::unordered_map<kv::ReconfigurationId, ByzantineIdentity> byids;
+
+  public:
+    ByzantineIdentitiesHook(
+      kv::Version version_, const ByzantineIdentities::Write& w) :
+      version(version_)
+    {
+      for (const auto& [rid, opt_bid] : w)
+      {
+        if (opt_bid.has_value())
+        {
+          LOG_DEBUG_FMT(
+            "New Byzantine network identity, valid for configuration #{}.",
+            rid);
+          byids.try_emplace(rid, opt_bid.value());
+        }
+      }
+    }
+
+    void call(kv::ConfigurableConsensus* consensus) override
+    {
+      for (auto& [rid, id] : byids)
+      {
+        consensus->add_identity(version, rid, id);
+      }
+    }
+  };
+
+  class IdentityTracker
+  {
+  public:
+    virtual void add_network_configuration(
+      const kv::NetworkConfiguration& config) = 0;
+    virtual void add_identity(
+      ccf::SeqNo seqno, kv::ReconfigurationId rid, ccf::Identity id) = 0;
+    virtual bool have_identity_for(
+      kv::ReconfigurationId rid, ccf::SeqNo commit_idx) const = 0;
+    virtual Identity get_identity(kv::ReconfigurationId rid) const = 0;
+    virtual void reshare(const kv::NetworkConfiguration& config) = 0;
+    virtual kv::ReconfigurationId find_reconfiguration(
+      const kv::Configuration::Nodes& nodes) const = 0;
+  };
+
+  class ByzantineIdentityTracker : public IdentityTracker
+  {
+  public:
+    enum class SessionState
+    {
+      STARTED,
+      FINISHED
+    };
+
+    class Session
+    {
+    public:
+      SessionState state = SessionState::STARTED;
+      kv::NetworkConfiguration config;
+
+      Session(const kv::NetworkConfiguration& config_) : config(config_) {}
+    };
+
+    ByzantineIdentityTracker(
+      std::shared_ptr<aft::State> shared_state_,
+      std::shared_ptr<enclave::RPCMap> rpc_map_,
+      crypto::KeyPairPtr node_sign_kp_,
+      const crypto::Pem& node_cert_) :
+      shared_state(shared_state_),
+      rpc_map(rpc_map_),
+      node_sign_kp(node_sign_kp_),
+      node_cert(node_cert_)
+    {}
+
+    virtual ~ByzantineIdentityTracker() {}
+
+    virtual void add_network_configuration(
+      const kv::NetworkConfiguration& config) override
+    {
+      network_configs[config.rid] = config;
+    }
+
+    virtual void reshare(const kv::NetworkConfiguration& config) override
+    {
+      auto rid = config.rid;
+      LOG_DEBUG_FMT("Identities: start resharing for configuration #{}", rid);
+      sessions.emplace(rid, Session(config));
+
+      auto msg = std::make_unique<threading::Tmsg<UpdateIdentityTaskMsg>>(
+        update_identity_cb, rid, rpc_map, node_sign_kp, node_cert);
+
+      threading::ThreadMessaging::thread_messaging.add_task(
+        threading::ThreadMessaging::get_execution_thread(
+          threading::MAIN_THREAD_ID),
+        std::move(msg));
+    }
+
+    virtual kv::ReconfigurationId find_reconfiguration(
```

Comment:
I would prefer an std::optional, that's what we do nearly everywhere in preference to sentinel values.